### ISSUE TITLE
Not working on TS 5.0

### ISSUE THREAD

**Issue by samchon** [cevek/ttypescript#140]
When run `ttsc` command, below error occurs:

```bash
D:\github\samchon\typia\node_modules\ttypescript\lib\patchCreateProgram.js:84
    tsm.createProgram = createProgram;
                      ^

TypeError: Cannot set property createProgram of #<Object> which has only a getter
    at patchCreateProgram (D:\github\samchon\typia\node_modules\ttypescript\lib\patchCreateProgram.js:84:23)
    at loadTypeScript (D:\github\samchon\typia\node_modules\ttypescript\lib\loadTypescript.js:21:56) 
    at Object.<anonymous> (D:\github\samchon\typia\node_modules\ttypescript\lib\tsc.js:8:46)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (D:\github\samchon\typia\node_modules\ttypescript\bin\tsc:2:1)
```

**Comment by szatanjl** [comment]
Don't know if it is related but I was looking through the code and I find the commit `f37f095` suspicious.

It adds check if tsc version is >= 4.9 (I believe that was the intent), but does so incorrectly `major >= 4 && minor >= 9` @ tsc.ts

**Comment by samchon** [comment]
You're right, I did a mistake on the version specification.

However, TS 5.0 breaks `ttypescript` too, even when fixing the version checking error.

**Comment by samchon** [comment]
This problem is based on TS 5.0 has changed target build option from `es5` to `es2018`.

I wrote an issue about that in TypeScript repo

https://github.com/microsoft/TypeScript/issues/52826

**Comment by jakebailey** [comment]
@samchon This problem has nothing to do with the target being changed. TypeScript is now itself authored in modules, which do not allow at-runtime modification of exports, as opposed to previous versions which used TS namespaces, which are just objects and could be modified (though any modification like that would be very, very fragile).

**Comment by nonara** [comment]
Here's a quick update for everyone - the changes in v5 required a new approach for a few reasons, but we have it solved and are about to release a new version of `ts-patch`.

We've opted to add drop-in replacement support for `ttypescript` which has had lagging maintenance support for a while now.

What this means is, you can install `ts-patch` and simply use `tspc`, the same as you would use `ttsc`. It will be on-the-fly, in memory modification, although it's notably faster due to our caching mechanism.

We're nearly finished, so you can expect this soon.

**Comment by sunrabbit123** [comment]
If you don't have time to migrate, enter
```sh
> yarn remove ttypescript
> yarn add -D ttsc
```

If so, your project will work just fine as is.
Unless you're only using ts-node and cache features.

If you have enough time to migrate, we recommend using ts-patch.

**Comment by nonara** [comment]
@sunrabbit123 last I saw, your solution lacks a critical piece, which will strip transformers of the full functionality that many will need.

Anyway, if you want to create another fork, that's up to you.

**Comment by sunrabbit123** [comment]
If so, it's simply a library that only works for now, so the
Isn't that a good thing for you, who only want one library?

I don't have a big picture, I just developed and deployed what I needed for the moment.
@nonara

**Comment by sunrabbit123** [comment]
In case you're wondering, the
I'm not interested in extending the functionality.
I'm just concerned with typescript version compatibility.

**Comment by sunrabbit123** [comment]
Numerous forks clutter the ecosystem, but they also make it mature.

Competition is necessary for better products,
As a company owner, I'm sure you know the benefits of having competition.

Of course, from the user's point of view, as you say, it's inconvenient and unsettling when the product, the interface, is not fixed.
That's what you're concerned about.

That's why I'm talking about it.
I'm not interested in extending features.

So any transpiler that works with TTSC will work with TSP.
Of course, it may not be the opposite.

**Comment by nonara** [comment]
So, one point here. I see that you're going out and opening PRs in major libraries (ts-jest). I also just looked at your code, and I can confirm that it isn't going to work. I'll detail below.

I think it's prudent for you to understand the scope of the problem and get your approach right before you start opening PRs. At the very least, let users know that it's not a complete solution.

Look at this file: https://github.com/sunrabbit123/ttypescript/blob/master/packages/ttypescript/src/tsc.ts#L15

I wrote the solution you see there to accommodate 4.9. That simply will not work with v5.

Look in https://unpkg.com/typescript@5.0.2/lib/tsc.js and search for "var StatisticType". It isn't in the file. Nor would it be.

If you want to confirm, run your code in the debugger and tell me if that replace does _anything_. If it doesn't, it won't work properly. If it does, it wouldn't even run, because it'd be broken due to the structural differences of v4 and v5.

Competition is fine — what bugs me is when someone is trying to inject an unworking solution into larger tools without an understanding of how it even works.

There is a reason Cevek and I have the lines we do regarding tsc. It's clear that you don't understand why we did it and that you don't understand that your code is broken. If you want to do this responsibly, I'd ask that you let people know that your fork is not complete and will break in _many_ situations, and please do not try to submit PRs to libraries until you've understood the problem and solved for it.

If you want to take it on, you can find some detailed explanations that I've written on what we're doing and why in the open issues on ts-patch, including a discussion with the MSFT engineer who made the changes in v5. 

Simply put, **please get a handle on the issue and submit a complete solution before trying to submit PRs to major tools that we all rely on.**

Or, you can simply wait for tsp, which is, as it's always been, backwards compatible with ttsc.

**Comment by sunrabbit123** [comment]
Thanks for the advice
But what the hell are you talking about
I don't understand your reference
>Look at this file: [sunrabbit123/ttypescript@master/packages/ttypescript/src/tsc.ts#L15](https://github.com/sunrabbit123/ttypescript/blob/master/packages/ttypescript/src/tsc.ts?rgh-link-date=2023-03-29T16%3A41%3A15Z#L15)
> I wrote the solution you see there to accommodate 4.9. That simply will not work with v5.
> Look in [unpkg.com/typescript@5.0.2/lib/tsc.js](https://unpkg.com/typescript@5.0.2/lib/tsc.js) and search for "var StatisticType". It isn't in the file. Nor would it be.

If you have a plugin that doesn't currently work with TTSC, please bring it up and tell us about it.
I'm too dumb to understand your reference
In 5.0.2, var StaticType is neither found nor touched

Ok, sir I'm v4 and my minor should be 9 or higher.
I have the wrong category

But that's not the immediate problem

**Comment by sunrabbit123** [comment]
Bring me a specific case of something not working
You have not properly proposed the code in question, nor the problem.

Please be precise and specific
At least that the test case is broken.

**Comment by nonara** [comment]
Hi all!

We have a beta ready for testing. It can directly replace ttypescript, using the same in-memory modification technique. Very easy to switch

Details here:

- https://github.com/nonara/ts-patch/issues/93#issuecomment-1500795063

Let me know if it works for you.

**Comment by mindplay-dk** [comment]
@nonara works for me! 👍
